### ISSUE TITLE
Support FIDO U2F

### ISSUE THREAD

**Issue by Unknwon** [gogits/gogs#4384]
Reference: https://github.com/tstranex/u2f

**Comment by bibo38** [comment]
I would like to work on this issue. Or is it already being worked on?

I would suggest/implement the [Web Authentication API](https://fidoalliance.org/fido2/fido2-web-authentication-webauthn/)(Webauthn), as it allows a broader range of two factor authentications instead of only U2F (e.g. [Windows Hello](https://docs.microsoft.com/en-us/microsoft-edge/dev-guide/windows-integration/web-authentication)).

The https://webauthn.io demo site lists the following go libraries to be used with WebAuthn:
- [duo_labs/webauthn](https://github.com/duo-labs/webauthn)
- [koesie10/webauthn](https://github.com/koesie10/webauthn)

I would choose the first library, as I don't know if the high level abstractions of `koesie10/webauthn` work well with the request routing in Gogs.

**Comment by bibo38** [comment]
> It would be great to have in Gogs, but also great effort to make it work :)

I agree. But it's a great opportunity to support Gogs and get (as a result) my most important missing feature I would like to see.

**Comment by unknwon** [comment]
@bibo38 agree, better to start with small, functioning, to fully featured.

**Comment by techy121592** [comment]
I don't mean to be bump an old issue, but this has become relevant to me so I was wondering what the status of this is.

**Comment by unknwon** [comment]
@techy121592 Status: I do not think anyone is actively working on this.

**Comment by bibo38** [comment]
Jep, thats correct... I've worked on it last year and functionality wise it is about 90% complete (i think the biggest issue was, that I didn't save the destination redirect after the WebAuthn login, it simply forwards you to the base url). On the code side it needs to be cleaned up (and the commits need to be squashed/reordered, currently they are just messy). And of course it needs to be updated to the current master.

Feel free to complete finish this issue based on my code: https://github.com/bibo38/gogs/tree/webauthn .
